tiledb/sm/c_api/tiledb.cc

```diff
@@ -4126,6 +4151,330 @@ int32_t tiledb_subarray_get_range_var_from_name(
   return TILEDB_OK;
 }
 
+/* ****************************** */
+/*      SUBARRAY PARTITIONER      */
+/* ****************************** */
+
+int32_t tiledb_subarray_partitioner_alloc(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    tiledb_subarray_partitioner_t** subarray_partitioner,
+    // TBD: require these here, or provide a separate set_memory_budget
+    // (subarray_partitioner::set_memory_budget() exists...)
+    uint64_t memory_budget,
+    uint64_t memory_budget_var,
+    uint64_t memory_budget_validity) {  //,
+  // Sanity check
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, subarray) == TILEDB_ERR)
+    return TILEDB_ERR;
+
+  // Create a buffer struct
+  *subarray_partitioner = new (std::nothrow) tiledb_subarray_partitioner_t;
+  if (*subarray_partitioner == nullptr) {
+    auto st = Status::Error("Failed to allocate TileDB subarray object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_OOM;
+  } else {
+    (*subarray_partitioner)->partitioner_ = nullptr;
+  }
+
+  // Create a new subarray object
+  try {
+    (*subarray_partitioner)->partitioner_ =
+        new (std::nothrow) tiledb::sm::SubarrayPartitioner(
+            &ctx->ctx_->storage_manager()->config(),
+            *subarray->subarray_,
+            memory_budget,
+            memory_budget_var,
+            memory_budget_validity,
+            ctx->ctx_->storage_manager()->compute_tp(),
+            subarray->subarray_->stats(),
+            ctx->ctx_->storage_manager()->logger());
+  } catch (...) {
+    // in case Subarray constructor (or involved sub-members) should throw.
+    //->partitioner_ already nullptr from above.
+  }
+  if ((*subarray_partitioner)->partitioner_ == nullptr) {
+    delete *subarray_partitioner;
+    auto st = Status::Error("Failed to allocate TileDB subarray object");
+    LOG_STATUS(st);
+    save_error(ctx, st);
+    return TILEDB_OOM;
+  }
+
+  // Success
+  return TILEDB_OK;
+}
+
+void tiledb_subarray_partitioner_free(
+    tiledb_subarray_partitioner_t** subarray_partitioner) {
+  if (subarray_partitioner != nullptr && *subarray_partitioner != nullptr) {
+    delete (*subarray_partitioner)->partitioner_;
+    delete (*subarray_partitioner);
+    *subarray_partitioner = nullptr;
+  }
+}
+
+int32_t tiledb_subarray_partitioner_set_layout(
+    tiledb_ctx_t* ctx,
+    tiledb_layout_t layout,
+    tiledb_subarray_partitioner_t* partitioner) {
+  // Sanity check
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, partitioner) == TILEDB_ERR)
+    return TILEDB_ERR;
+
+  // TBD: any validation on 'layout'?
```

Comment:
What kind of validation could we do here?

Reply:
Didn't know that's why I was asking, but in keeping with as little functionality in capi layer as possible, any validation (I now know) should be done within the Subarray being set, comment removed.